From 4ff25f3c1a1807f6ea79a911f6dc121a0f9b0559 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:34 -0400
Subject: [net] netfilter: nf_tables: check for unset NFTA_SET_ELEM_LIST_ELEMENTS attribute

Message-id: <20141021161934.22888.57242.stgit@localhost.localdomain>
Patchwork-id: 97981
O-Subject: [RHEL7 net PATCH 084/124] netfilter: nf_tables: check for unset NFTA_SET_ELEM_LIST_ELEMENTS attribute
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 7d5570ca8972aad58de0b4ad03a5e991c41d4ce2

 commit 7d5570ca8972aad58de0b4ad03a5e991c41d4ce2
 Author: Pablo Neira Ayuso <pablo@netfilter.org>
 Date:   Fri Jul 25 13:15:36 2014 +0200

    netfilter: nf_tables: check for unset NFTA_SET_ELEM_LIST_ELEMENTS attribute

    Otherwise, the kernel oopses in nla_for_each_nested when iterating over
    the unset attribute NFTA_SET_ELEM_LIST_ELEMENTS in the
    nf_tables_{new,del}setelem() path.

    netlink: 65524 bytes leftover after parsing attributes in process `nft'.
    [...]
    Oops: 0000 [#1] SMP
    [...]
    CPU: 2 PID: 6287 Comm: nft Not tainted 3.16.0-rc2+ #169
    RIP: 0010:[<ffffffffa0526e61>]  [<ffffffffa0526e61>] nf_tables_newsetelem+0x82/0xec [nf_tables]
    [...]
    Call Trace:
     [<ffffffffa05178c4>] nfnetlink_rcv+0x2e7/0x3d7 [nfnetlink]
     [<ffffffffa0517939>] ? nfnetlink_rcv+0x35c/0x3d7 [nfnetlink]
     [<ffffffff8137d300>] netlink_unicast+0xf8/0x17a
     [<ffffffff8137d6a5>] netlink_sendmsg+0x323/0x351
    [...]

    Fix this by returning -EINVAL if this attribute is not set, which
    doesn't make sense at all since those commands are there to add and to
    delete elements from the set.

    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index c12a90b..119de56 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3073,6 +3073,9 @@ static int nf_tables_newsetelem(struct sock *nlsk, struct sk_buff *skb,
  struct nft_ctx ctx;
  int rem, err = 0;
 
+ if (nla[NFTA_SET_ELEM_LIST_ELEMENTS] == NULL)
+  return -EINVAL;
+
  err = nft_ctx_init_from_elemattr(&ctx, skb, nlh, nla, true);
  if (err < 0)
   return err;
@@ -3156,6 +3159,9 @@ static int nf_tables_delsetelem(struct sock *nlsk, struct sk_buff *skb,
  struct nft_ctx ctx;
  int rem, err = 0;
 
+ if (nla[NFTA_SET_ELEM_LIST_ELEMENTS] == NULL)
+  return -EINVAL;
+
  err = nft_ctx_init_from_elemattr(&ctx, skb, nlh, nla, false);
  if (err < 0)
   return err;
-- 
1.7.1